---
title: "test enrichment"
output:
  html_document: default
  pdf_document: default
editor_options: 
  chunk_output_type: inline
---


```{r, message=FALSE, results='hide'}

# libraries

library(dplyr) 
library(data.table)
library(clusterProfiler) # for gseGO
library(enrichplot)
library(pathview)
library(org.Hs.eg.db)
library(stats) # for fisher test
library(rmarkdown)
library(ggnewscale) # for Visualization of Functional Enrichment Results
library(tidyr)
library(ReactomePA)
library(GSEABase)


```

# Exercise 1

Differential gene expression between two different
immune cell types, natural killer (NK) cells and CD4 T helper cells (Th)).
The data was generated by RNA sequencing and analyzed using limma.<br/>
Import file NK_vs_Th_diff_gene_exp.csv that contains gene ensembl ID,
gene symbols, log2 fold change, t-statistic and raw p-value


```{r}

# - import csv file NK_vs_Th_diff_gene_exercise_1.csv
path = '/Users/dianaavalos/Desktop/Enrichment_Analysis/Exercises'
NK_vs_Th = fread(
  file.path(path, 'NK_vs_Th_diff_gene_exercise_1.csv'),
  head = TRUE,
  stringsAsFactors = FALSE
)

# - create new column with adjusted p-values
NK_vs_Th$adjusted_pval <-
  p.adjust(NK_vs_Th$P.Value, method = "BH", n = length(NK_vs_Th$P.Value))
# p.adjust.methods # c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr", "none")

# - look for CPS1 gene in table
CPS1 = NK_vs_Th[which(NK_vs_Th$symbol == "CPS1")]
CPS1
```
### Answer the questions of Exercice 1

1. Is CPS1 significantly differentially expressed at alpha=0.05? <br/>
Yes, pval=0.044

2. How many genes are up-regulated and how many genes are down-regulated <br/>

```{r}

sum(NK_vs_Th$logFC < 0) # 8909
sum(NK_vs_Th$logFC > 0) # 11576
length(NK_vs_Th$t) # 20485

```
There are 8909 genes down-regulated and 11576 up-regulated <br/>

And after BH p-value adjustment?

```{r}
sum(NK_vs_Th$logFC < 0 & NK_vs_Th$adjusted_pval < 0.05) # 1895
sum(NK_vs_Th$logFC > 0 & NK_vs_Th$adjusted_pval < 0.05) # 1957

```

There are 1895 genes down-regulated and 1957 up-regulated after BH p-value adjustment <br/>

3. Is CPS1 still differentially expressed after p-value adjustment at alpha=0.05?

```{r}
CPS1 = NK_vs_Th[which(NK_vs_Th$symbol == "CPS1")]$adjusted_pval
print(CPS1)
```
No, adj pval=0.1565113




# Exercise 2


```{r, message=FALSE, results='hide'}

# - Import the list of genes belonging to the adaptive immune response
adimresp_term2gene = fread(
  file.path(path, 'adaptive_immune_response_geneset_term2gene.csv'),
  head = TRUE,
  stringsAsFactors = FALSE
)

```

- Is the adaptive immune response gene set significantly enriched in genes up-regulated in NK vs Th?

```{r}
# filter NK_vs_Th with genes belonging to the adaptive immune response
# Determine numbers of up-regulated genes that are within or outside of gene set:

nrow(adimresp_term2gene) #663

subset_signif <-
  NK_vs_Th %>% subset(NK_vs_Th$adjusted_pval < 0.05) 

subset_signif_upreg <-
  subset_signif %>% subset(subset_signif$logFC > 0 ) 

summary(subset_signif_upreg$symbol %in% adimresp_term2gene$V2) 

```
We have 3,852 significantly deferentially expressed genes (subset_signif)
We have 1,957  significantly up-regulated genes (subset_signif_upreg)
We have 75 significantly up-regulated genes belonging to the adaptive immune response
We have 1882 significantly up-regulated genes not belonging to the adaptive immune response

```{r}
subset_not_signif <-
  NK_vs_Th %>% subset(NK_vs_Th$adjusted_pval > 0.05 ) 

subset_not_significantly_upregulated <-
  NK_vs_Th %>% subset(NK_vs_Th$adjusted_pval > 0.05  |  NK_vs_Th$logFC < 0 ) 

summary(subset_not_signif$symbol %in% adimresp_term2gene$V2) 
summary(subset_not_significantly_upregulated$symbol %in% adimresp_term2gene$V2) 

```
I think there is an error in the solution: we want all the genes not significantly up-regulated meaning we want the genes not deferentially expressed but also the ones significantly down-regulated no?

```{r}
# create contingency table 
cont.table <- matrix(c(75, 270, 1882, 16363), ncol = 2, byrow = F)
cont.table_2nd_version <- matrix(c(75, 410, 1882, 18118), ncol = 2, byrow = F)

colnames(cont.table)<-c("up", "not_up")
rownames(cont.table)<-c("in_set", "not_in_set")
cont.table
```

```{r}
# - Run Fisher test of up-regulated genes
fisher.test(cont.table)
fisher.test(cont.table_2nd_version)
```

The adaptive immune response gene set is significantly enriched in genes up-regulated in NK vs Th as the p value is <0.01 

understanding odd's ratio:


https://towardsdatascience.com/interpretation-of-odds-ratio-and-fishers-exact-test-c6dde394d204#:~:text=It%20is%20called%20the%20Odds,greater%20or%20less%20than%201.

- How many GO gene sets are significant after GSEA (use minGSSize=30, pvalueCutoff=0.1) ?

```{r, message=FALSE, results='hide'}
# - Create named vector of t-statistics and run gsea of built-in GO gene sets:
idType(OrgDb = "org.Hs.eg.db") # to determine allowed gene id type, Genome wide annotation for Human.

gl<-NK_vs_Th$t
names(gl)<-make.names(NK_vs_Th$symbol, unique = T)
gl<-gl[order(gl, decreasing = T)]

# Gene set ontology calling
GO_NK_Th<-gseGO(gl, ont="BP",
                OrgDb = org.Hs.eg.db,
                keyType = "SYMBOL",
                minGSSize=30,
                pvalueCutoff = 0.1,
                seed=T)

```

```{r}
# To have a glance at the results table
length(GO_NK_Th@result$pvalue)
#head(GO_NK_Th@result)
```
There are 562 GO gene sets that are significant after GSEA for pvalueCutoff=0.1 ( length(GO_NK_Th@result$pvalue) ) <br/>

- Is the adaptive immune response gene set significant? Up-reg. or down-reg.?

```{r}
# Is the adaptive immune response gene set significant?
GO_NK_Th@result[GO_NK_Th@result$Description=="adaptive immune response",]
```
The adaptive immune response gene set is significant as the adjusted p-value is 8.35981e-08 <br/>

normalized enrichment score (NES) is negative (-1.849308) so this gene set is down regulated

- Are the majority of gene sets rather up-regulated or down-regulated?  <br/>

your list has to be ordered form the most upregulated to the most downregulated
so a positive NES for S1 will mean that genes over-represented in that set are upregulated in your dataset. (top of the list)  https://www.biostars.org/p/132575/

```{r}
# - Count the number of negative NES values and the number of positive NES values
summary(GO_NK_Th@result$NES > 0)
```
There are 460 gene sets rather up-regulated, and 102 down-regulated <br/>



#  Exercise 3
- First convert the gene symbols to EntrezID to perform a GSEA of KEGG
pathways (with argument minGSSize=30).

```{r, message=FALSE, results='hide'}

# - First convert the gene symbols to EntrezID 

library(org.Hs.eg.db)
hs <- org.Hs.eg.db
keytypes(org.Hs.eg.db)

# Use bitr to convert Ensembl IDs to ENTREZID and create a named vector of t-statistics
# to provide as gene list to the gseKEGG function, use minGSSize = 30
# NOTE: if the gsea fails, use readRDS to import the pre-computed results
# contained in the file gseKEGG_Nk_vs_Th_results.rds

gene_convert <- bitr(as.character(NK_vs_Th$ensembl_gene_id), 
                     fromType="ENSEMBL", 
                     toType=c("SYMBOL", "ENTREZID"), OrgDb="org.Hs.eg.db")

```

```{r}
head(gene_convert)
dim(gene_convert) # 16786
length(unique(gene_convert$ENTREZID)) # 16770

# GSEA of KEGG: create a vector of genes that are coded with the EntrezID:
# use the sorted gene list gl previously created:
gl_kegg<-cbind(SYMBOL=names(gl), t=gl)

# merge with converted gene symbols to combine both:
# by default the data frames are merged on the columns with names they both have
gl_kegg<-merge(gl_kegg, gene_convert)
head(gl_kegg)
```

```{r, message=FALSE, results='hide',warning=FALSE}

gl_kegg_list<-as.numeric(as.character(gl_kegg$t))
names(gl_kegg_list)<-as.character(gl_kegg$ENTREZID)
gl_kegg_list<-sort(gl_kegg_list, decreasing = T)

# run GSEA of KEGG:
KEGG_NK_Th<-gseKEGG(gl_kegg_list, organism = "hsa", "ncbi-geneid",
                    minGSSize = 30,
                    eps=1e-50,
                    seed=T)

```

Are the majority of gene sets rather up-regulated or down-regulated?
```{r}
summary(KEGG_NK_Th@result$NES > 0)
```
There are 19 gene sets rather up-regulated, and 7 down-regulated <br/>


Is there a KEGG immune-related gene set coming up? Is there a KEGG Natural
 killer gene set coming up?
```{r}
subset(KEGG_NK_Th@result,grepl("immune", KEGG_NK_Th@result$Description))
subset(KEGG_NK_Th@result,grepl("killer", KEGG_NK_Th@result$Description))
```
There are no KEGG immune-related gene set coming up, but there is 1 KEGG Natural
killer gene set coming up <br/>


```{r}
# How many built-in KEGG pathways?
length(KEGG_NK_Th@geneSets)
```
There are 344 KEGG pathways in our diferentially enriched gene set <br/>

If you want to see which genes are included in one of the built-in KEGG pathways,
 where could you find this information?
 
```{r}
subset(KEGG_NK_Th@result,grepl("killer",KEGG_NK_Th@result$Description))$core_enrichment

genes_in_the_killer_cell_pathway= KEGG_NK_Th@geneSets$hsa04650
genes_ID_all_names_killer_cell  <- gl_kegg %>% subset(gl_kegg$ENTREZID %in% genes_in_the_killer_cell_pathway ) 
head(genes_ID_all_names_killer_cell)
```

Import the hallmark gene sets and run a GSEA. How many significant gene sets are there?
```{r}
hallmark_gene_sets<-read.gmt(file.path(path, 'h.all.v7.1.symbols.gmt'))
head(hallmark_gene_sets)
```

```{r, message=FALSE, results='hide',warning=FALSE}
length(unique(hallmark_gene_sets$term))

# Run GSEA with the function that allows to use custom gene sets, 
# provide the named vector of t statistics
h_NK_vs_Th<-GSEA(gl, TERM2GENE = hallmark_gene_sets,
                eps=1e-50, 
                seed=T)
# what is eps? @param eps This parameter sets the boundary for calculating the p value.

# head(h_NK_vs_Th@result)
```

```{r}
# Number of significant gene sets:
length(which(h_NK_vs_Th@result$p.adjust<=0.05)) 
```
there are 4 out of 11 significant gene sets

```{r}
# A quick dotplot of the hallmark results:
dotplot(h_NK_vs_Th, orderBy="p.adjust")

```
# Exercise 4

Create figures for the GSEA results:
- barplot of -log10(p-value) of top 10 GO p-values

```{r}
par(mar=c(5,20,3,3)) 
barplot(rev(-log10(GO_NK_Th@result$pvalue[1:10])),
        horiz = T, names=rev(GO_NK_Th@result$Description[1:10]),
        las=2, xlab="-log10(adj.p-value)",
        cex.names = 0.7,
        col="aquamarine3") 
abline(v=-log10(0.05))
```
 barcode plot of one of the significant hallmark gene sets (choose one)
# TODO
```{r}
gseaplot(h_NK_vs_Th, geneSetID = "HALLMARK_MTORC1_SIGNALING",
         title="HALLMARK_MTORC1_SIGNALING")
```

```{r}
h_NK_vs_Th@result$Description[1:4]
gseaplot(h_NK_vs_Th, geneSetID = "HALLMARK_MYC_TARGETS_V2",
         title="HALLMARK_MYC_TARGETS_V2")
```
```{r}
h_NK_vs_Th@result$Description[1:4]
gseaplot(h_NK_vs_Th, geneSetID = "HALLMARK_OXIDATIVE_PHOSPHORYLATION",
         title="HALLMARK_OXIDATIVE_PHOSPHORYLATION")
```

```{r}
#cnetplot(h_NK_vs_Th, categorySize="pvalue", foldChange=geneList)
```

pathview map for KEGG Natural Killer mediated cytotoxicity (optional: with none-significant genes in grey)

```{r, message=FALSE, results='hide',warning=FALSE}
subset_killer=subset(KEGG_NK_Th@result, grepl("killer", KEGG_NK_Th@result$Description))

# pathview with non-significant genes in grey:
# set log fold change of non-significant genes to 0:
NK_vs_Th$logFC_0<-ifelse(NK_vs_Th$adjusted_pval>0.05, 0, NK_vs_Th$logFC)

# create named vector of fold change values:
genePW<-NK_vs_Th$logFC_0
names(genePW)<-NK_vs_Th$symbol
genePW<-na.omit(genePW)

# create pathview map of Natural killer cell mediated cytotoxicity = hsa04650
pathview(gene.data  = genePW,
         pathway.id = "hsa04650",
         species    = "hsa",
         gene.idtype = "SYMBOL")

```

Barplot of NES of the top GO gene sets, with different color for up- or down-reg gene sets:
```{r}

sorted_GO_NK_Th<- GO_NK_Th@result[order(GO_NK_Th@result$NES, decreasing = F),]
sorted_GO_NK_Th$color<-ifelse(sorted_GO_NK_Th$NES<0, "red", "darkblue")

par(mar=c(5,20,3,3))
barplot(sorted_GO_NK_Th$NES[c(1:10, (nrow(sorted_GO_NK_Th)-9):nrow(sorted_GO_NK_Th))],
        horiz = T, names=sorted_GO_NK_Th$Description[c(1:10, (nrow(sorted_GO_NK_Th)-9):nrow(sorted_GO_NK_Th))],
        las=2, xlab="NES",
        cex.names = 0.7,
        col=sorted_GO_NK_Th$color[c(1:10, (nrow(sorted_GO_NK_Th)-9):nrow(sorted_GO_NK_Th))]) 

```
```{r}
NK_up_2 <-subset(NK_vs_Th, 
              NK_vs_Th$adjusted_pval<=0.05&NK_vs_Th$logFC>3)
NK_up_2<-as.character(NK_up_2$symbol)
```

```{r}
ego <- enrichGO(NK_up_2, OrgDb='org.Hs.eg.db', ont="BP", keyType = "SYMBOL")
```
```{r}
barplot(ego)
```
```{r}
dotplot(ego, orderBy="p.adjust")
```
Which genes are shared by several gene sets:
```{r}
cnetplot(ego, categorySize="pvalue")
```

```{r}
# Similarity among gene sets:
ego2 <- pairwise_termsim(ego)
emapplot(ego2)
emapplot_cluster(ego2)

```
#  Extra exercise for credits

Perform GSEA of the NK vs Th data using the Reactome gene sets downloaded on the 
MSigDB website.

- How many gene sets are significantly enriched? Generate an ordered barplot of
the NES of all genesets, and generate a barcode plot for the gene set with the lowest NES
Steps

Use GSEA providing a sorted named vector of t-statistics and the reactome gene sets,
use minGSSize=30
NOTE: if the GSEA fails, use readRDS to import the pre-computed results contained in the
file "reactomeGSEA_NK_vs_Th_results.rds"
- count the number of significant adjusted p-values
- use barplot() and gseaplot() for the visualization of the results



```{r}
# Perform GSEA of the NK vs Th data using the Reactome gene sets downloaded on the MSigDB website.
# download  reactome pathways and gene sets associated

reactome_database <- 
  read.table(
    file.path(path, 'c2.cp.reactome.v7.1.symbols.gmt'),
    sep = "\t",
    fill = TRUE,
    header = F,
  )

# - import the reactome gene sets in the file c2.cp.reactome.v7.1.symbols.gmt using read.gmt() , this function doesnt work and shows only first gene of pathway
reactome_gmt <- read.gmt(file.path(path, 'c2.cp.reactome.v7.1.symbols.gmt'))
```


```{r, message=FALSE, results='hide',warning=FALSE}
# formating
reactome_database_except_col_1_2 = subset(reactome_database, select = -c(1,2) )

genes_reac_pathway <- reactome_database_except_col_1_2 %>%
  unite("genes",sep=", ")

reactome_database_bis <- merge(reactome_database[,1:2],genes_reac_pathway)

# # i don't know how to integrate this in a function...
# reactome_NK_vs_Th<-GSEA(gl, TERM2GENE = reactome_database_bis,
#                  eps=1e-50, 
#                  seed=T)
```
other option:
```{r, message=FALSE, results='hide',warning=FALSE}
# https://bioc.ism.ac.jp/packages/3.2/bioc/vignettes/ReactomePA/inst/doc/ReactomePA.html
gl_kegg_list<-as.numeric(as.character(gl_kegg$t))
names(gl_kegg_list)<-as.character(gl_kegg$ENTREZID)
gl_kegg_list<-sort(gl_kegg_list, decreasing = T)
# gl_kegg_list is the list of genes with ENTREZID symbols in decreasing order

reactome <-
  gsePathway(
    gene = gl_kegg_list,
    nPerm = 1000,
    pvalueCutoff = 0.2,
    pAdjustMethod = "BH",
    verbose = FALSE
  )

# enrichPathway also uses reactome ontology
x <- enrichPathway(gene=gene_convert$ENTREZID,pvalueCutoff=0.05, readable=T)
head(summary(x))

```

- How many gene sets are significantly enriched? <br/>
 sum(reactome@result$p.adjust<0.05)<br/>
[1] 0<br/>
 sum(reactome@result$p.adjust>0.05)<br/>
[1] 89<br/>


for reactome:<br/>
```{r}

sorted_reactome_NK_Th<- reactome@result[order(reactome@result$NES, decreasing = F),]
sorted_reactome_NK_Th$color<-ifelse(sorted_reactome_NK_Th$NES<0, "red", "darkblue")

par(mar=c(5,20,3,3))
barplot(sorted_reactome_NK_Th$NES[c(1:10, (nrow(sorted_reactome_NK_Th)-9):nrow(sorted_reactome_NK_Th))],
        horiz = T, names=sorted_reactome_NK_Th$Description[c(1:10, (nrow(sorted_reactome_NK_Th)-9):nrow(sorted_reactome_NK_Th))],
        las=2, xlab="NES",
        cex.names = 0.7,
        col=sorted_reactome_NK_Th$color[c(1:10, (nrow(sorted_reactome_NK_Th)-9):nrow(sorted_reactome_NK_Th))]) 

```
for KEGG
```{r}

sorted_KEGG_NK_Th<- KEGG_NK_Th@result[order(KEGG_NK_Th@result$NES, decreasing = F),]
sorted_KEGG_NK_Th$color<-ifelse(sorted_KEGG_NK_Th$NES<0, "red", "darkblue")

par(mar=c(5,20,3,3))
barplot(sorted_KEGG_NK_Th$NES[c(1:10, (nrow(sorted_KEGG_NK_Th)-9):nrow(sorted_KEGG_NK_Th))],
        horiz = T, names=sorted_KEGG_NK_Th$Description[c(1:10, (nrow(sorted_KEGG_NK_Th)-9):nrow(sorted_KEGG_NK_Th))],
        las=2, xlab="NES",
        cex.names = 0.7,
        col=sorted_KEGG_NK_Th$color[c(1:10, (nrow(sorted_KEGG_NK_Th)-9):nrow(sorted_KEGG_NK_Th))]) 


```
Generate an ordered barplot of the NES of all genesets, 
and generate a barcode plot for the gene set with the lowest NES <br/>

Barplot of NES of the top reactome gene sets, with different color for up- or down-reg gene sets:<br/>
```{r}
sorted_GO_NK_Th$set='GO'
sorted_KEGG_NK_Th$set='KEGG'
sorted_reactome_NK_Th$set='REAC'

all_genesets <- bind_rows(sorted_GO_NK_Th,sorted_KEGG_NK_Th,sorted_reactome_NK_Th)
all_genesets<- all_genesets[order(all_genesets$NES, decreasing = F),]

```

```{r}
pdf(paste0(path,"/all_genesets_NES.pdf"),         # File name
    width = 8, height = 7, # Width and height in inches
    bg = "white",          # Background color
    paper = "A4")          # Paper size

LENGTH=length(all_genesets$NES)
#LENGTH=100

barplot(
  all_genesets$NES[c(1:LENGTH, (nrow(all_genesets) - 9):nrow(all_genesets))],
  horiz = T,
  names = all_genesets$Description[c(1:LENGTH, (nrow(all_genesets) - 9):nrow(all_genesets))],
  las = 2,
  xlab = "NES",
  cex.names = 0.1,
  col = all_genesets$color[c(1:LENGTH, (nrow(all_genesets) - 9):nrow(all_genesets))]
)

dev.off()
```